token-erc-20/chaincode-java/src/main/java/org/example/TokenERC20Contract.java

```diff
@@ -0,0 +1,519 @@
+
+        // Check if the transferred value is less than the allowance
+        if (currentAllowance < valueInt) {
+
+            String errorMessage = String.format("The spender does not have enough allowance to spend.");
+            System.out.println(errorMessage);
+            throw new ChaincodeException(errorMessage);
+
+        }
+
+        boolean transferResp = this.doTransfer(ctx, from, to, valueInt);
+
+        if (!transferResp) {
+            throw new ChaincodeException("Failed to transfer");
+        }
+
+        // Decrease the allowance
+        long updatedAllowance = currentAllowance - valueInt;
+        stub.putStringState(allowanceKey.toString().trim(), String.valueOf(updatedAllowance));
+        System.out.printf("spender %s allowance updated from %d to %d", spender, currentAllowance, updatedAllowance);
+
+        JSONObject obj = new JSONObject();
+        obj.put("from", from);
+        obj.put("to", to);
+        obj.put("value", valueInt);
+        stub.setEvent("Transfer", this.serialize(obj));
+        System.out.println("transferFrom ended successfully");
+
+        return true;
+    }
+
+    @Transaction()
+    private boolean doTransfer(final Context ctx, final String _from, final String _to, long _value) {
+
+        if (_from.equalsIgnoreCase(_to)) {
+            throw new ChaincodeException("cannot transfer to and from same client account");
+        }
+
+        if (_value < 0) { // transfer of 0 is allowed in ERC20, so just validate against negative amounts
+            throw new ChaincodeException("transfer amount cannot be negative");
+        }
+
+        ChaincodeStub stub = ctx.getStub();
+        // Retrieve the current balance of the sender
+        CompositeKey fromBalanceKey = stub.createCompositeKey(balancePrefix, _from.trim());
+        String fromCurrentBalance = stub.getStringState(fromBalanceKey.toString().trim());
+        if (fromCurrentBalance.isBlank() || fromCurrentBalance.length() == 0) {
+            String errorMessage = String.format("client account %s has no balance", _from);
+            throw new ChaincodeException(errorMessage);
+
+        }
+
+        long _fromCurrentBalance = Long.parseLong(fromCurrentBalance.toString().trim());
+
+        // Check if the sender has enough tokens to spend.
+        if (_fromCurrentBalance < _value) {
+            String errorMessage = String.format("client account %s has insufficient funds", _from);
+            throw new ChaincodeException(errorMessage);
+        }
+
+        // Retrieve the current balance of the recepient
+        CompositeKey toBalanceKey = stub.createCompositeKey(balancePrefix, _to);
+        String toCurrentBalance = stub.getStringState(toBalanceKey.toString().trim());
+
+        long _toCurrentBalance = 0;
+        // If recipient current balance doesn't yet exist, we'll create it with a
+        // current balance of 0
+        if (toCurrentBalance.isBlank() || toCurrentBalance.length() == 0) {
```

Comment:
Second check is unreachable as it duplicates the first.
```suggestion
        if (toCurrentBalance.isBlank()) {
```